executable-spec/src/Cardano/Ledger/Spec/STS/Update/Data.hs

```diff
@@ -1,29 +1,130 @@
+{-# LANGUAGE DeriveAnyClass #-}
 {-# LANGUAGE DeriveGeneric #-}
 {-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE FlexibleInstances #-}
 
 module Cardano.Ledger.Spec.STS.Update.Data where
 
-import Data.Word (Word64)
+--import Data.Word (Word64)
 import           Data.Monoid.Generic (GenericMonoid (GenericMonoid),
                      GenericSemigroup (GenericSemigroup))
 import           Data.Set (Set)
 import           GHC.Generics (Generic)
-
+import           Data.Text (Text)
+import           Numeric.Natural (Natural)
+import           Data.Hashable (Hashable)
+import qualified Data.Hashable as H
 import qualified Ledger.Core as Core
+import           Data.Map.Strict (Map)
+
+
+-- |A unique ID of a software update
+newtype UpId = UpId { getUpId :: Integer --Word64
+                    }
+  deriving (Show, Eq, Ord, Hashable, Generic)
+
+instance Core.HasHash (UpId) where
+  hash (UpId i) = Core.Hash $ H.hash i
+
+instance Num UpId where
+  (+) a b = UpId $ (getUpId a) + (getUpId b)
+
+-- | Protocol version
+data ProtVer = ProtVer
+  { _pvMaj :: Natural
+  , _pvMin :: Natural
+  , _pvAlt :: Natural
+  } deriving (Eq, Generic, Ord, Show, Hashable)
+
+-- | Application version
+newtype ApVer = ApVer Natural
+  deriving stock (Generic, Show)
+  deriving newtype (Eq, Ord, Num, Hashable)
+
+-- | Consensus Protocol Parameter Name
+data ParamName =
+    BlockSizeMax
+  | TxSizeMax
+  | SlotSize
+  | EpochSize
+  deriving (Eq, Generic, Ord, Show, Hashable)
+
+-- | Metadata structure for SIP
+data SIPMetadata =
+  SIPMetadata { versionFrom :: (ProtVer, ApVer)
+                -- ^ The version the this SIP has been based on
+              , versionTo :: (ProtVer, ApVer)
+              -- ^ the version after the SIP takes effect
+              , impactsConsensus :: Bool
```

Comment:
Maybe here we could explain what `True` and `False` mean, or even use an algebraic data type to express more clearly what the possible impact on consensus is.

Reply:
OK I will do it.

Reply:
I have created a new ADT for this.